📝 Testing in a federated arch (#152)
apollographql/docs: src/content/technotes/TN0007-testing-in-a-federated-architecture-unit-tests.mdx

---
title: Testing in a Federated Architecture
tags: [server, router, testing]
---

> Testing in a federated architecture has three general parts: unit testing, integration testing, and composition testing.

Reviewer: To add on to this- we should also be clarifying what this is attempting to test; is this testing a subgraph, gateway, or both?

> Testing in a federated architecture has three general parts: unit testing, integration testing, and composition testing.

Reviewer: Last note on this- we should clarify this is Apollo Server/Gateway specific; other subgraph libraries will be tested much differently for unit tests specifically.

> Testing in a federated architecture has three general parts: unit testing, integration testing, and composition testing.

Author: PTAL when pushed (I'll hit the request review button again)

## Unit testing

> We recommend unit testing all your resolvers. When unit testing your resolvers, we recommend mocking as much of the data as possible using something like [`@faker-js/faker`](https://www.npmjs.com/package/@faker-js/faker). This package allows you to use realistic fake data when mocking inputs and outputs.

Reviewer: ```suggestion
We recommend unit testing all your resolvers. When unit testing your resolvers, we recommend mocking as much of the data as possible using a package like [`@faker-js/faker`](https://www.npmjs.com/package/@faker-js/faker). This package allows you to use realistic fake data when mocking inputs and outputs.
```

> We recommend unit testing all your resolvers. When unit testing your resolvers, we recommend mocking as much of the data as possible using something like [`@faker-js/faker`](https://www.npmjs.com/package/@faker-js/faker). This package allows you to use realistic fake data when mocking inputs and outputs.

Author: added an example, don't know if it was what you were looking for though. PTAL

### `__resolveReference`

> A [`__resolveReference`](https://www.apollographql.com/docs/federation/entities/#2-define-a-reference-resolver) warrants a specific call out in a federated architecture. A `__resolveReference` is how subgraphs communicate, making testing this code very important.

Reviewer: ```suggestion
A [`__resolveReference`](/federation/entities/#2-define-a-reference-resolver) resolver warrants a specific callout in a federated architecture. A `__resolveReference` resolver is how subgraphs resolve entities across subgraphs, making testing this code very important.
```

Same as above- code samples can help illustrate here.

Author: would a code sample really help here? the test would look very similar to how a normal resolvers test looks. the only real difference is there is no args param.

## Integration testing

When doing integration tests, please consider the following:

- Operations should be pulled from current live queries
  - Our recommended way is to look at the operations currently going through your graph and using those operations for testing.

>   - Current queries can be pulled from Apollo studio's API.

Reviewer: To add a couple thoughts: 

- The operations should be representative of actual queries if you aren't currently live (e.g. building new resolvers/queries) 
- Example of the API call would be nice as well, as it could be a way to populate it automatically for more thorough testing; maybe worth an apollosolutions repo?

Author: updated the section with a call out to that - PTAL when pushed (I'll hit the request review button again)

>   - Current queries can be pulled from Apollo studio's API.

Author: They do i was just thinking about removing these when looking at the comment above.

- Random test queries/mutations should be avoided
  - Random tests end up testing paths that are not realistic. We suggest pulling the queries from Apollo studio instead of using arbitrary queries.

> - Cardinality should be considered

Reviewer 2: ```suggestion
- Use variables to ensure high operation cardinality.
  - If your test operations don't use any GraphQL variables (or if you use the _same_ variable values across executions), your supergraph is likely to return cached data. This circumvents large portions of execution logic, limiting test effectiveness.
  - By using a variety of operations and variable values, you help make sure that your tests result in minimal cache hits.
```

>   - When running tests, specifically integration tests, low cardinality tests hit caches. The best way to avoid caches is to have a high cardinality in the variables used. We recommend mitigating the problem by using as many variables in your queries/mutations as possible. This allows for easy changing of inputs.

Reviewer 2: ```suggestion
```

> ### `__resolveReference` and integrations testing subgraphs

Reviewer: ```suggestion
### `__resolveReference` and integration testing subgraphs
```

> ### `__resolveReference` and integrations testing subgraphs

Reviewer 2: ```suggestion
### Reference resolvers in integration tests
```

> Depending on your schema, you may not start from the query node in a subgraph. The integration tests in this situation involves mimicking the gateway/router. You can execute a query against `_entities` to do integration tests. This looks like the following:

Reviewer: ```suggestion
Depending on your schema, you may not start from the query node in a subgraph. The integration tests in this situation involves mimicking the gateway/router. You can execute [a query against `_entities`](/federation/federation-spec/#resolve-requests-for-entities) to do integration tests. This looks like the following:
```

> Depending on your schema, you may not start from the query node in a subgraph. The integration tests in this situation involves mimicking the gateway/router. You can execute a query against `_entities` to do integration tests. This looks like the following:

Reviewer 2: ```suggestion
If your supergraph schema includes federated [entities](/federation/entities), you should create integration tests that mimic your gateway/router executing the reference resolvers for those entities. This is done by querying the special `Query._entities` field that's automatically present in every subgraph schema:
```

```js
const query = `query GetEntities($representations: [_Any!]!) {
  _entities(representations: $representations) {
   ... on User {
     firstName
   }
  }
}`;
```

> 

Reviewer 2: ```suggestion

For more information, see [`Query._entities`](/federation/subgraphs#query_entities).

```

## Composition testing

> Composition testing is specific to federated architectures. It involves testing that your graph will compose and handle the queries that are currently sent to it by clients. This is handled through [rover](https://www.apollographql.com/docs/rover/). These checks are done by running the [`rover subgraph check`](https://www.apollographql.com/docs/rover/commands/subgraphs#checking-subgraph-schema-changes) command. This is best in your CI/CD pipeline.

Reviewer: ```suggestion
Composition testing is specific to federated architectures. It involves testing that your graph will compose and handle the queries that are currently sent to it by clients. This is handled through [Rover](/docs/rover/). These checks are done by running the [`rover subgraph check`](/rover/commands/subgraphs#checking-subgraph-schema-changes) command. This is best used when configured within your CI/CD pipeline.
```

> Composition testing is specific to federated architectures. It involves testing that your graph will compose and handle the queries that are currently sent to it by clients. This is handled through [rover](https://www.apollographql.com/docs/rover/). These checks are done by running the [`rover subgraph check`](https://www.apollographql.com/docs/rover/commands/subgraphs#checking-subgraph-schema-changes) command. This is best in your CI/CD pipeline.

Reviewer 2: ```suggestion
Composition testing is specific to a federated architecture. It involves testing that your subgraph schemas successfully [compose into a supergraph schema](/federation/federated-types/composition/) that can resolve the operations sent by clients. You perform these tests with the [Rover CLI](/rover/), via the [`rover subgraph check`](/rover/commands/subgraphs#subgraph-check) command. We recommend performing this check in your CI/CD pipeline.
```
